### PR TITLE
Allow setting of kubeRbacProxyMainResources in kube-state-metrics

### PR DESCRIPTION
**Pull request by Luis-TT**

### CODE REVIEW

**Comment by ArthurSens** [comment]
kube-rbac-proxy resources can be set by editing its own library, isn't that enough?

https://github.com/prometheus-operator/kube-prometheus/blob/ae12388b33954fe228423dc2898719bdec2c0006/jsonnet/kube-prometheus/components/kube-rbac-proxy.libsonnet#L1-L39

**Comment by Luis-TT** [comment]
I tried setting it like this:
`local kp =
  (import 'kube-prometheus/main.libsonnet') +
  // Uncomment the following imports to enable its patches
  // (import 'kube-prometheus/addons/anti-affinity.libsonnet') +
  // (import 'kube-prometheus/addons/managed-cluster.libsonnet') +
  // (import 'kube-prometheus/addons/node-ports.libsonnet') +
  // (import 'kube-prometheus/addons/static-etcd.libsonnet') +
  // (import 'kube-prometheus/addons/custom-metrics.libsonnet') +
  // (import 'kube-prometheus/addons/external-metrics.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
      kubeRbacProxy+: {
        resources+: {
          limits+: { cpu: '400m' },
        },
      },
    },
  };`
and
`local kp =
  (import 'kube-prometheus/main.libsonnet') +
  // Uncomment the following imports to enable its patches
  // (import 'kube-prometheus/addons/anti-affinity.libsonnet') +
  // (import 'kube-prometheus/addons/managed-cluster.libsonnet') +
  // (import 'kube-prometheus/addons/node-ports.libsonnet') +
  // (import 'kube-prometheus/addons/static-etcd.libsonnet') +
  // (import 'kube-prometheus/addons/custom-metrics.libsonnet') +
  // (import 'kube-prometheus/addons/external-metrics.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
    kubeRbacProxy+: {
        resources+: {
          limits+: { cpu: '400m' },
        },
      },
  };`

both didn't work.

I think maybe because the arguments are overriden by the components creating the kube-rbac-proxy
https://github.com/prometheus-operator/kube-prometheus/blob/ae12388b33954fe228423dc2898719bdec2c0006/jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet#L95

**Review by paulfantom — CHANGES_REQUESTED** [review]
I think we made a mistake by providing `kubeRbacProxyImage` parameter as it paved a way to add parameters like one from this PR. We should probably expose a configuration option like `kubeRbacProxy` to allow setting all configuration parameters that are in https://github.com/prometheus-operator/kube-prometheus/blob/main/jsonnet/kube-prometheus/components/kube-rbac-proxy.libsonnet#L1-L39 instead of mapping them one by one.

That said I am against introducing new parameters that are only mapping particular settings from kube-rbac-proxy. This complicates the codebase quite a lot while not having too many benefits. Especially since `resources` can be overridden on the higher level.

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 1]
```diff
@@ -11,7 +11,21 @@ local defaults = {
...
-
+  kubeRbacProxyMain: {
+    resources+: {
+      limits+: { cpu: '40m' },
+      requests+: { cpu: '20m' },
+    },
+    limits+: { cpu: '40m' },
+    requests+: { cpu: '20m' },
+    name: 'kube-rbac-proxy-main',
+    upstream: 'http://127.0.0.1:8081/',
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:22]: This cannot be in `defaults` and need to be overridden in L102.

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 2]
```diff
@@ -11,7 +11,21 @@ local defaults = {
...
   },
-
+  kubeRbacProxyMain: {
+    resources+: {
+      limits+: { cpu: '40m' },
+      requests+: { cpu: '20m' },
+    },
+    limits+: { cpu: '40m' },
+    requests+: { cpu: '20m' },
+    name: 'kube-rbac-proxy-main',
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:21]: This cannot be in `defaults` and need to be overridden in L102.

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 3]
```diff
@@ -11,7 +11,21 @@ local defaults = {
...
+    upstream: 'http://127.0.0.1:8081/',
+    secureListenAddress: ':8443',
+    ports: [
+      { name: 'https-main', containerPort: 8443 },
+    ],
+    image: $.kubeRbacProxyImage,
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:27]: Due to how jsonnet can override things, we should set this in L102 and not here.

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 4]
```diff
@@ -11,7 +11,21 @@ local defaults = {
...
+      requests+: { cpu: '20m' },
+    },
+    limits+: { cpu: '40m' },
+    requests+: { cpu: '20m' },
+    name: 'kube-rbac-proxy-main',
+    upstream: 'http://127.0.0.1:8081/',
+    secureListenAddress: ':8443',
+    ports: [
+      { name: 'https-main', containerPort: 8443 },
+    ],
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:26]: This cannot be in `defaults` and need to be overridden in L102.

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 5]
```diff
@@ -11,7 +11,21 @@ local defaults = {
...
+  kubeRbacProxyMain: {
+    resources+: {
+      limits+: { cpu: '40m' },
+      requests+: { cpu: '20m' },
+    },
+    limits+: { cpu: '40m' },
+    requests+: { cpu: '20m' },
+    name: 'kube-rbac-proxy-main',
+    upstream: 'http://127.0.0.1:8081/',
+    secureListenAddress: ':8443',
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:23]: This cannot be in `defaults` and need to be overridden in L102.

**Review by paulfantom — CHANGES_REQUESTED** [review]
I like the direction, but we cannot have everything in `defaults` as this leaves too much room for breaking the stack. For example, there is no benefit of overriding `upstream` for kube-rbac-proxy and as such, this setting shouldn't be in `defaults.kubeRbacProxyMain`, but should be set in `local kubeRbacProxyMain = krp(ksm._config.kubeRbacProxyMain)`

Ideally, we should get sth like:
```
local kubeRbacProxyMain = krp(ksm._config.kubeRbacProxyMain) + {
  upstream: 'http://127.0.0.1:8081/',
  // other non-overridable, critical settings here.
},
```

File: jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet [hunk 6]
```diff
@@ -12,6 +12,12 @@ local defaults = {
     limits: { cpu: '100m', memory: '250Mi' },
   },
 
+  kubeRbacProxyMain: {
+    resources+: {
+      limits+: { cpu: '40m' },
+      requests+: { cpu: '20m' },
+    },
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:19]: L101 to L104 should be removed as they are overriding what is set here.

Comment (Luis-TT) [reply jsonnet/kube-prometheus/components/kube-state-metrics.libsonnet:19]: sorry forget to remove L101 to L104 :grimacing:

**Review by paulfantom — APPROVED** [review]
lgtm

**Pull request merged**